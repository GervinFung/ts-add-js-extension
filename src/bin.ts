#!/usr/bin/env node
import main from './';
import { a } from './a.mjs';
console.log({ a });
main(process.argv.join(' '));
