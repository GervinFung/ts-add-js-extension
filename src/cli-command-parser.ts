import fs from 'fs';
import path from 'path';
import { guard } from './type';
import pkg from './package';

const commandKeyWords = {
	help: {
		isMandatory: false,
		keyword: 'help',
	},
	version: {
		isMandatory: false,
		keyword: 'version',
	},
	dir: {
		isMandatory: true,
		keyword: 'dir',
	},
	include: {
		isMandatory: false,
		keyword: 'include',
	},
	showChanges: {
		type: 'deprecated',
		isMandatory: false,
		keyword: 'showchanges',
		reason: 'The function showchanges is deprecated in favor of `showprorgess`',
	},
	assignment: {
		assign: '=',
		key: '--',
	},
} as const;

class TokenParser {
	constructor(private readonly token: Token) {}

	readonly parseVersion = () => {
		const { keyword } = commandKeyWords.version;
		if (this.token.key !== keyword) {
			return {
				exists: false,
			} as const;
		}
		return {
			type: keyword,
			exists: true,
			value: pkg.version,
		} as const;
	};

	readonly parseHelp = () => {
		const { keyword } = commandKeyWords.help;
		if (this.token.key !== commandKeyWords.help.keyword) {
			return {
				exists: false,
			} as const;
		}
		return {
			type: keyword,
			exists: true,
			value: fs.readFileSync('public/help.md', { encoding: 'utf-8' }),
		} as const;
	};

	readonly parseDir = () => {
		const { keyword } = commandKeyWords.dir;
		if (this.token.key !== commandKeyWords.dir.keyword) {
			return {
				exists: false,
			} as const;
		}

		return {
			type: keyword,
			exists: true,
			value: guard({
				value: this.token.value.split(' ').at(0),
				error: new Error(
					'There must be at least one element in values for dir'
				),
			}),
		} as const;
	};

	readonly parseInclude = () => {
		const { keyword } = commandKeyWords.include;
		if (this.token.key !== commandKeyWords.include.keyword) {
			return {
				exists: false,
			} as const;
		}

		return {
			type: keyword,
			exists: true,
			value: this.token.value.split(' '),
		} as const;
	};

	readonly parseShowChanges = () => {
		const { keyword } = commandKeyWords.showChanges;
		const { key, value } = this.token;
		if (key !== commandKeyWords.showChanges.keyword) {
			return {
				exists: false,
			} as const;
		}

		if (value === 'true' || value === 'false') {
			return {
				type: keyword,
				exists: true,
				value: JSON.parse(value) as boolean,
			} as const;
		}

		throw new Error(
			`${key}=${value} is invalid, it can only receive boolean value`
		);
	};

	readonly processNonRecognisableToken = () => {
		return {
			type: 'invalid',
			reason: 'then token cannot be recognized',
			token: this.token,
		} as const;
	};
}

type Args = ReadonlyArray<string>;

type Token = Readonly<{
	key: string;
	value: string;
}>;

export default class ParseArgs {
	private readonly tokens: ReadonlyArray<Token>;

	private constructor(args: Args) {
		this.tokens = this.tokenize(args);
	}

	static readonly create = (arg: string) => {
		const tokens = arg.split(commandKeyWords.assignment.key);
		const name = ParseArgs.checkPackageName(tokens.at(0));
		if (name !== 'proceed') {
			throw name.error;
		}

		if (tokens.includes('add')) {
			console.log(
				'The "add" in the command can be removed, as it is only used for backward compatibility'
			);
		}
		return new this(
			tokens
				.map((token) => (token === 'add' ? '' : token))
				.filter((_, index) => index)
		);
	};

	private static readonly checkPackageName = (name: string | undefined) =>
		name?.includes(pkg.name ?? '')
			? 'proceed'
			: {
					error: new Error(
						`The pkg name "${name}" passed is invalid`
					),
			  };

	private readonly tokenize = (args: Args) => {
		const { assign } = commandKeyWords.assignment;

		return args
			.flatMap((arg) => {
				if (arg.includes(assign)) {
					return [arg];
				}

				const [key, ...value] = arg.split(' ');
				return [`${key}${assign}${value.join(' ')}`];
			})
			.map((args): Token => {
				const [key, value] = args.split(assign);

				return {
					key: guard({
						value: key,
						error: new Error(
							'Key cannot be undefined after being flat mapped'
						),
					}),
					value: guard({
						value,
						error: new Error(
							'Value cannot be undefined after being flat mapped'
						),
					}).trim(),
				};
			});
	};

	readonly asVersion = () =>
		this.tokens.reduce(
			(result, token) => {
				if (result.exists) {
					return result;
				}
				return new TokenParser(token).parseVersion();
			},
			{ exists: false } as ReturnType<TokenParser['parseVersion']>
		);

	readonly asHelp = (): ReturnType<TokenParser['parseHelp']> => {
		if (!this.tokens.length) {
			return {
				type: 'help',
				exists: true,
				value: fs.readFileSync(path.join('public', 'help.md'), {
					encoding: 'utf-8',
				}),
			};
		}
		return this.tokens.reduce(
			(result, token) => {
				if (result.exists) {
					return result;
				}
				return new TokenParser(token).parseHelp();
			},
			{ exists: false } as ReturnType<TokenParser['parseHelp']>
		);
	};

	readonly asOperation = () => {
		const processedToken = this.tokens.map((token) => {
			const parser = new TokenParser(token);
			const dir = parser.parseDir();
			if (dir.exists) {
				return dir;
			}
			const include = parser.parseInclude();
			if (include.exists) {
				return include;
			}
			const showChanges = parser.parseShowChanges();
			if (showChanges.exists) {
				return showChanges;
			}
			return parser.processNonRecognisableToken();
		});

		processedToken
			.flatMap((node) => (node.type !== 'invalid' ? [] : [node]))
			.forEach((node) =>
				console.log(
					`The "${node.token}" in the command is invalid as ${node.reason}. So please remove it`
				)
			);

		const nodes = processedToken.flatMap((node) =>
			node.type === 'invalid' ? [] : [node]
		);

		return {
			dir: guard({
				value: nodes.find((node) => node.type === 'dir')
					?.value as ReturnType<TokenParser['parseDir']>['value'],
				error: new Error(
					'dir is a mandatory field, it should be present to know which dir it should operate on'
				),
			}),
			// optional
			include: nodes.find((node) => node.type === 'include')
				?.value as ReturnType<TokenParser['parseInclude']>['value'],
			showChanges: nodes.find((node) => node.type === 'showchanges')
				?.value as ReturnType<TokenParser['parseShowChanges']>['value'],
		} as const;
	};
}

type TrueConfig = ReturnType<ParseArgs['asOperation']>;

type PartialConfig = Partial<Omit<TrueConfig, 'dir'>> & Pick<TrueConfig, 'dir'>;

export type { TrueConfig, PartialConfig };
