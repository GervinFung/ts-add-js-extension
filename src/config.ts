import type { PartialConfig } from './cli-command-parser';

/**
 * @deprecated since version 1.4.0
 * Will be deleted in version 2.0
 * There is no need to parse config with this function
 * As configurations can be passed directly
 * Using this function will halt the program
 * And requires you to write configs directly
 * */
const parseConfig = (_: Readonly<Record<string, unknown>>) => {
    throw new Error(
        [
            `Function parseConfig should not be used, this function exists because of "yargs"`,
            `Now yargs is removed, cli arguments parser is hand-written`,
        ].join('\n')
    );
};

const valuelizeConfig = (config: PartialConfig) =>
    ({
        ...config,
        showChanges: config.showChanges ?? true,
        include: config.include ?? [],
    } as const);

export { valuelizeConfig, parseConfig };
export type { PartialConfig as ParsedConfig };
