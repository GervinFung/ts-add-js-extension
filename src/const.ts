const separator = '/';

const extensions = {
	javaScript: ['.js', '.mjs', '.jsx'],
	typeDefinition: ['.d.ts', '.d.mts'],
} as const;

const matchJs = (filePath: string) => {
	return extensions.javaScript.find((extension) => {
		return filePath.endsWith(extension);
	});
};

const matchDts = (filePath: string) => {
	return extensions.typeDefinition.find((extension) => {
		return filePath.endsWith(extension);
	});
};

const matchEither = (filePath: string) => {
	return matchJs(filePath) ?? matchDts(filePath);
};

export { extensions, matchJs, matchDts, matchEither, separator };
