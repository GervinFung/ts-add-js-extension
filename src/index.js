const start = ({ size, index }) => {
    const percentage = Math.round((index / size) * 100);
    const line = [
        `Updating file: src/process/pgtyped/hash/${index}.ts`,
        `Total file${!size ? '' : 's'}: ${size}\tDone: ${index}\tPercentage: ${
            size === index
                ? 100
                : !((percentage * 10) % 10)
                ? percentage
                : percentage.toFixed(2)
        }\n`,
    ].join('\n');

    if (index !== 1) {
        process.stdout.moveCursor(0, -2);
        process.stdout.clearScreenDown();
    }

    process.stdout.write(line);
};

Array.from({ length: 100 }, (_, index) => {
    start({
        index: index + 1,
        size: 100,
    });
});
