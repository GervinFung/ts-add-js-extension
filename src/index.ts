import * as Eslint from '@typescript-eslint/typescript-estree';
import fs from 'fs';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';
import { Table } from 'console-table-printer';
import path from 'path';

type Node = Readonly<{
    file: string;
    code: string;
    ast: Eslint.AST<
        Readonly<{
            loc: true;
        }>
    >;
}>;

type AddJSNode = Readonly<{
    before: string;
    after: string;
}>;

type WriteCode = Readonly<{
    code: string;
    file: string;
}>;

type Files = ReadonlyArray<string>;

type ReplaceNodes = ReadonlyArray<AddJSNode>;

type Dir = string;

const getAllJavaScriptFiles = (dir: string): Files =>
    fs.readdirSync(dir).flatMap((file) => {
        const filePath = `${dir}/${file}`;
        return fs.statSync(filePath).isDirectory()
            ? getAllJavaScriptFiles(filePath)
            : !(path.extname(filePath) === '.js')
            ? []
            : [filePath];
    });

const readCode = (files: string): Promise<string> =>
    new Promise((resolve, reject) => {
        let fetchData = '';
        fs.createReadStream(files)
            .on('data', (data) => {
                fetchData = data.toString();
            })
            .on('end', () => resolve(fetchData))
            .on('error', reject);
    });

const getAllJavaScriptCodes = (files: Files): ReadonlyArray<Promise<Node>> =>
    files.map(async (file) => {
        const code = await readCode(file);
        return {
            file,
            code,
            ast: Eslint.parse(code, { loc: true }),
        };
    });

const formProperFilePath = ({
    delimiter,
    filePath,
}: Readonly<{
    delimiter: string;
    filePath: string;
}>) => filePath.split(delimiter).filter(Boolean).join(delimiter);

const addJSExtension = ({
    filePath,
    delimiter,
    importPath,
}: Readonly<{
    filePath: string;
    importPath: string;
    delimiter: string;
}>): Readonly<
    | {
          type: 'skip';
      }
    | {
          type: 'proceed';
          importPath: string;
          filePathForFileImported: string;
      }
> => {
    const isDirectory =
        fs.existsSync(filePath) && fs.lstatSync(filePath).isDirectory();
    const isJavaScript = !isDirectory && path.extname(filePath) === '.js';
    if (isJavaScript) {
        return {
            type: 'skip',
        };
    }
    const defaultFile = `${!isDirectory ? '' : '/index'}.js`;
    return {
        type: 'proceed',
        importPath: formProperFilePath({
            delimiter,
            filePath: `${importPath}${defaultFile}`,
        }),
        filePathForFileImported: formProperFilePath({
            delimiter,
            filePath: `${filePath}${defaultFile}`,
        }),
    };
};

const importExportWithJSExtension = ({
    files,
    node: {
        ast: { body },
        code,
        file,
    },
}: Readonly<{
    node: Node;
    files: Files;
}>): ReadonlyArray<
    WriteCode &
        Readonly<{
            replaceNodes: ReplaceNodes;
        }>
> => {
    const codeWithoutCarriageReturn = code.replace(/\r/gm, '');
    const splitted = codeWithoutCarriageReturn.split('\n');
    const statementType = Eslint.AST_NODE_TYPES;

    const replaceNodes: ReplaceNodes = body.flatMap((statement) => {
        const { type } = statement;
        switch (type) {
            case statementType.ImportDeclaration:
            case statementType.ExportAllDeclaration:
            case statementType.ExportNamedDeclaration: {
                const { source } = statement;
                if (!source) {
                    return [];
                } else {
                    const {
                        value,
                        loc: { end },
                    } = source;
                    const delimiter = '/';
                    const fileName = formProperFilePath({
                        delimiter,
                        filePath: !value.endsWith(delimiter)
                            ? value
                            : value.slice(0, -1),
                    })
                        .split(delimiter)
                        .pop();
                    if (!fileName) {
                        throw new Error(
                            `Impossible for file name to be non-existent for ${value}`
                        );
                    }
                    if (!value.startsWith('.')) {
                        return [];
                    } else {
                        const result = addJSExtension({
                            delimiter,
                            importPath: value,
                            filePath: path.join(file, '..', value),
                        });

                        switch (result.type) {
                            case 'skip':
                                return [];
                            case 'proceed': {
                                // if file name not included in list of js file read
                                const { filePathForFileImported, importPath } =
                                    result;
                                const fileFound = files.find((file) =>
                                    file.endsWith(filePathForFileImported)
                                );
                                if (!fileFound) {
                                    return [];
                                }

                                const before = splitted[end.line - 1];
                                if (!before) {
                                    throw new Error(
                                        `Old Code: ${before} is undefined`
                                    );
                                }
                                return [
                                    {
                                        before,
                                        after: before.replace(
                                            value,
                                            importPath
                                        ),
                                    },
                                ];
                            }
                        }
                    }
                }
            }
            default:
                return [];
        }
    });
    return [
        {
            file,
            replaceNodes,
            code: !replaceNodes.length
                ? codeWithoutCarriageReturn
                : replaceNodes.reduce(
                      (prev, { before, after }) => prev.replace(before, after),
                      codeWithoutCarriageReturn
                  ),
        },
    ];
};

const tabulateChanges = (
    changes: ReadonlyArray<{
        replaceNodes: ReplaceNodes;
        file: string;
    }>
) => {
    const table = new Table({
        columns: [
            { name: 'index', alignment: 'left' },
            { name: 'file', alignment: 'left' },
            { name: 'before', alignment: 'left' },
            { name: 'after', alignment: 'left' },
        ],
    });
    table.table.title = 'The Output of ts-add-js-extension';
    changes.forEach(({ replaceNodes, file }, changeIndex) =>
        replaceNodes.forEach(({ before, after }, index) => {
            const currentIndex = index + 1;
            table.addRow(
                {
                    file,
                    before,
                    after,
                    index:
                        changes
                            .map((change, index) =>
                                index >= changeIndex
                                    ? 0
                                    : change.replaceNodes.length
                            )
                            .reduce((prev, curr) => prev + curr) + currentIndex,
                },
                { color: 'cyan' }
            );
        })
    );
    table.printTable();
};

const postProcess = async ({
    dir,
    shouldLog,
    showChanges,
    withJSExtension,
}: Readonly<{
    dir: Dir;
    shouldLog: boolean;
    showChanges: boolean;
    withJSExtension: ReturnType<typeof importExportWithJSExtension>;
}>) => {
    await Promise.all(
        withJSExtension.map(({ code, file }) =>
            fs.writeFile(file, code, (err) => {
                if (err) {
                    console.error(err);
                }
            })
        )
    );
    if (showChanges) {
        tabulateChanges(
            withJSExtension.map(({ replaceNodes, file }) => ({
                file,
                replaceNodes: replaceNodes.flat(),
            }))
        );
    }
    if (withJSExtension.length && shouldLog && dir) {
        console.log(
            'Completed adding .js extension to each relative import/export statement'
        );
    }
};

const preProcess = ({
    dir,
    shouldLog,
}: Readonly<{
    dir: Dir;
    shouldLog: boolean;
}>) => {
    const files = getAllJavaScriptFiles(dir);
    if (shouldLog && !files.length) {
        console.log(
            `No files with .js extension was found in the specified folder of ${dir}. If this behavior is unexpected, Please file an issue, your feedback is greatly appreciated. Adios...`
        );
    }
    if (shouldLog && dir) {
        console.log(
            'Adding .js extension to each relative import/export. Please be patient...'
        );
    }
    return files;
};

const midProcess = async ({
    files,
    include,
}: Readonly<{
    include: ReadonlyArray<string>;
    files: ReturnType<typeof getAllJavaScriptFiles>;
}>) => {
    // user may import files from `common` into `src`
    const allIncludedFiles: Files = files.concat(
        include.flatMap(getAllJavaScriptFiles)
    );
    return (
        await getAllJavaScriptCodes(files).reduce(
            async (prev, curr) => (await prev).concat(await curr),
            Promise.resolve([] as ReadonlyArray<Node>)
        )
    ).flatMap((node) =>
        importExportWithJSExtension({
            node,
            files: allIncludedFiles,
        })
    );
};

type Argv = Parameters<
    Parameters<ReturnType<typeof yargs>['command']>[0]['handler']
>[0];

type ParsedConfig = Readonly<{
    dir: string;
    showChanges?: boolean;
    include?: ReadonlyArray<string>;
}>;

const parseConfig = (argv: Argv): ParsedConfig => {
    const include = argv['include'];
    return {
        dir: argv['dir'] as string,
        showChanges: argv['showchanges'] as boolean,
        include: !Array.isArray(include)
            ? undefined
            : include.map((dir) => {
                  const type = typeof dir;
                  if (type === 'string') {
                      return dir as string;
                  }
                  throw new Error(
                      `expect ${dir} to be string, got dir: ${dir} as ${type} instead`
                  );
              }),
    };
};

const finalizedConfig = (config: ParsedConfig) =>
    ({
        ...config,
        showChanges: config.showChanges ?? false,
        include: config.include ?? [],
    } as const);

const tsAddJsExtension = async (parsedConfig: ParsedConfig) => {
    const config = finalizedConfig(parsedConfig);
    const shouldLog = process.env.TS_ADD_JS_EXTENSION_NODE_ENV !== 'test';
    return postProcess({
        shouldLog,
        dir: config.dir,
        showChanges: config.showChanges,
        withJSExtension: await midProcess({
            include: config.include,
            files: preProcess({
                shouldLog,
                dir: config.dir,
            }),
        }),
    })
        .then(
            () =>
                ({
                    type: 'done',
                } as const)
        )
        .catch(
            (error) =>
                ({
                    type: 'error',
                    error,
                } as const)
        );
};

const main = (args: Array<string>) => {
    const describe =
        'Use to add .js extension for the relative import/export statement in the JavaScript code if there is lack of .js extension in the import/export statement.';
    return yargs(hideBin(args))
        .usage(describe)
        .command({
            command: 'add',
            describe,
            builder: {
                dir: {
                    type: 'string',
                    demandOption: true,
                    describe: 'The folder that need to add .js extension',
                },
                include: {
                    type: 'array',
                    demandOption: false,
                    describe:
                        'The folder of files that is imported or included in `dir` folder, exclusing the `dir` specified',
                },
                showchanges: {
                    default: true,
                    type: 'boolean',
                    demandOption: false,
                    describe:
                        'Show changes made to import/export declaration in table format',
                },
            },
            handler: (argv) => {
                tsAddJsExtension(parseConfig(argv))
                    .then((result) => {
                        switch (result.type) {
                            case 'error':
                                console.dir(
                                    {
                                        error: result.error,
                                    },
                                    { depth: null }
                                );
                        }
                    })
                    .catch((error) => console.dir({ error }, { depth: null }));
            },
        })
        .example(
            "Assume javascript files are placed in folder called 'build'\nThe command will be as below\n$0 add --dir=dist --include=common dist build --showchanges=true",
            `.
            1. "dir" stands for the directory of that needs to add .js extension. (string)
            2. "include" stands for the directory of files that is imported or included in 'dir' folder, exclusing the 'dir' specified. (array)
            3. "showchanges" determines whether to show the changes made to import/export in table format. (boolean)
            `
        )
        .help()
        .strict().argv;
};

export { parseConfig, tsAddJsExtension };
export default main;
