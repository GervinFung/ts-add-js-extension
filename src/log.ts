export default class Log {
	static readonly fromNumberOfFiles = (numberOfFiles: number) =>
		new Log(numberOfFiles, 0);

	private constructor(
		private readonly numberOfFiles: number,
		private completedFiles: number
	) {
		if (!this.numberOfFiles) {
			throw new Error(
				'Progress cannot be instantiated when there is no files to changes'
			);
		}
	}

	// ref: https://chrisyeh96.github.io/2020/03/28/terminal-colors.html
	private readonly redify = (word: string) => `\x1b[31m${word}\x1b[0m`;

	private readonly cyanify = (word: string) => `\x1b[36m${word}\x1b[0m`;

	private readonly boldify = (word: string) => `\x1b[1:37m${word}\x1b[0m`;

	readonly increment = ({
		file,
		repeat,
		succeed,
	}: Readonly<{
		file: string;
		repeat: number;
		succeed: boolean;
	}>) => {
		if (this.numberOfFiles <= this.completedFiles) {
			throw new Error(
				`Number of files succeed: ${this.completedFiles} cannot be the same as number of files: ${this.numberOfFiles}`
			);
		} else {
			this.completedFiles++;
			console.log(
				`${this.completedFiles}. ${' '.repeat(
					this.numberOfFiles.toString().length -
						this.completedFiles.toString().length
				)}${this.boldify(file)}${' '.repeat(repeat - file.length)} - ${
					succeed ? this.cyanify('SUCCEED') : this.redify('FAILED')
				}`
			);
		}
	};

	readonly end = ({
		errors,
	}: Readonly<{
		errors: ReadonlyArray<
			Readonly<{
				file: string;
				error: NodeJS.ErrnoException;
			}>
		>;
	}>) => {
		if (errors.length) {
			console.error(
				`The following file${
					errors.length === 1 ? '' : 's'
				} failed to add either .js or .mjs file extension`
			);
			console.error(
				Array.from(errors)
					.sort((a, b) => a.file.length - b.file.length)
					.map(
						({ file, error }, index) =>
							`${index}. file: ${file}, reason: ${error.message}`
					)
					.join('\n')
			);
		}
	};
}
