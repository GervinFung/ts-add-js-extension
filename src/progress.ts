import readline from 'readline';

export default class Progress {
    private numberOfFilesRan: number;
    private numberOfFilesCompleted: number;

    static readonly fromNumberOfFiles = (numberOfFiles: number) =>
        new Progress(numberOfFiles);

    private constructor(private readonly numberOfFiles: number) {
        if (!this.numberOfFiles) {
            throw new Error(
                'Progress cannot be instantiated when there is no files to changes'
            );
        }
        this.numberOfFilesRan = 0;
        this.numberOfFilesCompleted = 0;
    }

    readonly incrementNumberOfFilesRan = () => {
        if (this.numberOfFiles > this.numberOfFilesRan) {
            this.numberOfFilesRan++;
        } else {
            throw new Error(
                `number of files ran: ${this.numberOfFilesRan} cannot be the same as number of files: ${this.numberOfFiles}`
            );
        }
    };

    readonly incrementNumberOfFilesCompleted = () => {
        if (this.numberOfFiles > this.numberOfFilesCompleted) {
            this.numberOfFilesCompleted++;
        } else {
            throw new Error(
                `number of files completed: ${this.numberOfFilesCompleted} cannot be the same as number of files: ${this.numberOfFiles}`
            );
        }
    };

    private readonly cyanify = (number: string | number) =>
        `\x1b[36m${number}\x1b[0m`;

    private readonly progressLog = (
        params: Readonly<
            | {
                  percentage: number;
              } & (
                  | {
                        status: 'completed';
                    }
                  | {
                        status: 'ongoing';
                        filePath: string;
                    }
              )
        >
    ) => {
        const { percentage } = params;
        const diff = this.numberOfFilesRan - this.numberOfFilesCompleted;
        return [
            `Progress: total file${
                !this.numberOfFiles ? '' : 's'
            }: ${this.cyanify(this.numberOfFiles)}`,
            `resolved: ${this.cyanify(this.numberOfFilesCompleted)}`,
            `${!diff ? '' : `unresolved: ${this.cyanify(diff)}`}`,
            `percentage: ${this.cyanify(
                !((percentage * 10) % 10) ? percentage : percentage.toFixed(2)
            )}`,
            `${
                params.status === 'completed'
                    ? ''
                    : `resolved: ${this.cyanify(params.filePath)}`
            }`,
        ]
            .filter(Boolean)
            .join(', ');
    };

    readonly end = ({
        errors,
    }: Readonly<{
        errors: ReadonlyArray<
            Readonly<{
                file: string;
                error: NodeJS.ErrnoException;
            }>
        >;
    }>) => {
        console.log(
            this.progressLog({
                status: 'completed',
                percentage: Math.round(
                    (this.numberOfFilesCompleted / this.numberOfFiles) * 100
                ),
            })
        );

        if (errors.length) {
            console.error(
                `The following file${
                    errors.length === 1 ? '' : 's'
                } failed to add either .js or .mjs file extension`
            );
            console.error(
                Array.from(errors)
                    .sort((a, b) => a.file.length - b.file.length)
                    .map(
                        ({ file, error }) =>
                            `file: ${file}, reason: ${error.message}`
                    )
                    .join('\n')
            );
        }
    };

    readonly show = (filePath: string) => {
        this.incrementNumberOfFilesCompleted();
        const { numberOfFiles, numberOfFilesCompleted } = this;
        const params =
            numberOfFiles === numberOfFilesCompleted
                ? ({
                      status: 'completed',
                  } as const)
                : ({
                      status: 'ongoing',
                      filePath,
                  } as const);

        if (numberOfFilesCompleted === 1) {
            process.stdout.write('\n');
        } else {
            readline.moveCursor(
                process.stdout,
                0,
                numberOfFilesCompleted > 2 ? -1 : -2
            );
            readline.clearScreenDown(process.stdout);
        }

        if (numberOfFiles > numberOfFilesCompleted) {
            const line = this.progressLog({
                ...params,
                percentage: Math.round(
                    (numberOfFilesCompleted / numberOfFiles) * 100
                ),
            });
            process.stdout.write(`${line}\n`);
        }
    };
}
