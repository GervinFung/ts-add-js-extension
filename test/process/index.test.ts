import fs from 'fs';
import path from 'path';

import { describe, it, expect } from 'vitest';

import { tsAddJsExtension } from '../../src';

describe('ts add js extension', () => {
	const getPath = (subPath: string) => {
		return path.join(__dirname, subPath);
	};

	const getAllActualCodeWithFilePath = (
		dir: string
	): ReadonlyArray<
		Readonly<{
			code: Readonly<{
				actual: () => string;
				expected: () => string;
			}>;
		}>
	> => {
		return fs.readdirSync(dir).flatMap((file) => {
			const filePath = path.join(dir, file);

			if (fs.statSync(filePath).isDirectory()) {
				return getAllActualCodeWithFilePath(filePath);
			}

			return [
				{
					code: {
						actual: () => {
							return fs.readFileSync(filePath, {
								encoding: 'utf-8',
							});
						},
						expected: () => {
							return fs.readFileSync(
								filePath.replace('actual', 'expected'),
								{
									encoding: 'utf-8',
								}
							);
						},
					},
				},
			];
		});
	};

	describe('for JavaScript files only', () => {
		const parentDir = getPath(path.join('actual-result', 'js'));

		const javaScriptIncludes = fs
			.readdirSync(parentDir)
			.map((childPath) => {
				return path.join(parentDir, childPath);
			});

		describe.each(javaScriptIncludes)(
			'assert that it will work for JavaScript files with import/export statement',
			(dir) => {
				const result = tsAddJsExtension({
					config: {
						dir,
						include: javaScriptIncludes,
					},
				});

				it('should be able to append either js or mjs file extension', async () => {
					expect((await result).type).toBe('done');
				});

				it.each(getAllActualCodeWithFilePath(dir))(
					'should assert that file extension was added to proper import/export statement for file %s',
					async ({ code }) => {
						await result;
						expect(code.actual()).toBe(code.expected());
					}
				);
			}
		);
	});

	describe('for Type Definition files only', () => {
		const parentDir = getPath(path.join('actual-result', 'dts'));

		describe.each(
			fs.readdirSync(parentDir).map((childPath) => {
				return path.join(parentDir, childPath);
			})
		)(
			'assert that it will work for Type Definition files with or without JavaScript',
			(dir) => {
				const result = tsAddJsExtension({
					config: {
						dir,
					},
				});

				it(`should be able to append .js/.mjs extension for Type Definition file of ${dir}`, async () => {
					expect((await result).type).toBe('done');
				});

				it.each(getAllActualCodeWithFilePath(dir))(
					'should assert that file extension was added to proper import/export statement',
					async ({ code }) => {
						await result;
						expect(code.actual()).toBe(code.expected());
					}
				);
			}
		);
	});
});
