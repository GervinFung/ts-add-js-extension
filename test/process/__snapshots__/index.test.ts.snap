// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ts add js extension > should read the code and ensure each import/export statemnt is properly formed for "js/1.js" 1`] = `
"import * as fs from 'fs';
import abcdefgihijk, {
    box,
    dialog,
    table,
    cell,
    div,
    h1,
    h2,
    h4,
    h5,
    h6,
} from './2.js';
import { fs } from './2.js';
import * as sample2 from './2.js';
import * as sampleExport2 from '../js/1.js';
import * as index from './index.js';
import * as index from './index.js';
export { index } from './index.js';
export * as index from './index.js';
export { default as x } from './index.js';
export { default as x } from './index.js';
"
`;

exports[`ts add js extension > should read the code and ensure each import/export statemnt is properly formed for "js/2.js" 1`] = `
"import { sample1 } from './1.js';
import './c.css';
export * from './1.js';
export * from '../js/1.js';
"
`;

exports[`ts add js extension > should read the code and ensure each import/export statemnt is properly formed for "js/3.js" 1`] = `
"import * as sampleExport2 from './2.js';
export * from './1.js';
export * as ays from './1.js';
"
`;

exports[`ts add js extension > should read the code and ensure each import/export statemnt is properly formed for "js/4.mjs" 1`] = `
"import * as sampleExport1 from './1.js';
import * as sampleExport2 from './2.js';
export * from './1.js';
export * as ays from './1.js';
export * as abc from './6.mjs';
"
`;

exports[`ts add js extension > should read the code and ensure each import/export statemnt is properly formed for "js/5.js" 1`] = `
"import * as sampleExport3 from './3.js';
import * as sampleExport4 from './4.mjs';
"
`;

exports[`ts add js extension > should read the code and ensure each import/export statemnt is properly formed for "js/6.mjs" 1`] = `""`;

exports[`ts add js extension > should read the code and ensure each import/export statemnt is properly formed for "js/index.js" 1`] = `
"console.log('hi');
"
`;

exports[`ts add js extension > should read the code and ensure each import/export statemnt is properly formed for "ts/dist/index.d.ts" 1`] = `
"export * from './utils/index.js';
"
`;
